### PR TITLE
client-go: make transport cacheable using Interfaces instead of callback functions

### LINKED ISSUE #111911: Using exec auth in kubectl triggers one connection per applied/described object
### What happened?

Switching from the gcp auth provider to the exec auth provider in 1.25 made `kubectl apply` requests which previously made a single connection to the API server make one connection per applied object.

If enough items are applied, the connections can start to fail because of ephemeral port exhaustion

/kind bug
/milestone v1.25
/sig auth cli api-machinery

### What did you expect to happen?

Switching to the exec auth plugin would allow commands that previously worked to continue to work

### How can we reproduce it (as minimally and precisely as possible)?

Reproducer integration test at https://github.com/liggitt/kubernetes/commits/exec-auth-reproducer demonstrating one transport (and therefore a new connection) for each item in an applied file when using exec auth

on that branch:
```
go test ./test/integration/client/ -run TestApplyMultipleItemsWithExecPlugin -v
```
and observe 10 transport creations to apply 10 items (plus a couple others for discovery):
```
before apply
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
creating transport
after apply
```

### Anything else we need to know?

This happens because exec auth setup unconditionally sets the GetCert field in transport config here (even if the plugin is only going to return token credentials):

https://github.com/kubernetes/kubernetes/blob/58c10aa6eb5adfb1f3aa4d6cb898b8c347ba9e72/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L303

That makes the transport config uncacheable here:

https://github.com/kubernetes/kubernetes/blob/58c10aa6eb5adfb1f3aa4d6cb898b8c347ba9e72/staging/src/k8s.io/client-go/transport/cache.go#L141-L144

Commands like `kubectl describe` and `kubectl apply` repeatedly construct transports when dealing with multiple objects. This poor behavior was masked in past configurations by the transports being cached.

### Kubernetes version

tried with 1.24 and 1.25, reproduced with both

### Cloud provider

n/a

### OS version

n/a

### Install tools

n/a

### Container runtime (CRI) and version (if applicable)

n/a

### Related plugins (CNI, CSI, ...) and versions (if applicable)

n/a

### PR DESCRIPTION
**Pull request by aojea**
/kind feature
/kind api-change
/kind deprecation

#### What this PR does / why we need it:

The client-go constructors generates the Transport used by the clients,  and caches it when possible.

However, when the users need to add custom callbacks to the configuration, specifically the options

https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/rest/config.go#L133-L134

https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/rest/config.go#L140-L141

That are cascaded to the transport config

https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/transport/config.go#L70-L78

that also has another callback

https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/transport/config.go#L146

This causes that transport is not cacheable when one of this option is set

https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/transport/cache.go#L141-L143

and this can cause that clients generate a lot of connections for the same destination, this is especially problematic for `kubectl` , that generates one transport for each type, causing the exhausing of TCP port easily.

Some changes were introduced to fix this problem https://github.com/kubernetes/kubernetes/pull/105490, but until all users of client-go move to that approach, we need to be able to cache as much as possible.

This PR adds 3 new configuration options based on Interfaces to deprecate the current callbacks on the configuration options.

#### Which issue(s) this PR fixes:
Fixes #111911

#### Special notes for your reviewer:

This is based on the assumption that interfaces can be used as key of maps https://go.dev/play/p/PgBOb_wF2nH

We should have some expert on the area verifying this assumption

#### Does this PR introduce a user-facing change?
```release-note
client-go rest and transport: add two new interfaces types on both packages:

type Dialer interface {
	DialContext(ctx context.Context, network, address string) (net.Conn, error)
}

type Proxier interface {
	Proxy(*http.Request) (*url.URL, error)
}

client-go rest: Config deprecate the options
- Dial func(ctx context.Context, network, address string) (net.Conn, error)
- Proxy func(*http.Request) (*url.URL, error)
 
in favor of two new options that implement the new interfaces:
- Dialer Dialer
- Proxier Proxier

client-go transport: add a new interface
type Certificate interface {
	GetCertificate() (*tls.Certificate, error)
}

client-go transport: TLSConfig deprecate the `GetCert func() (*tls.Certificate, error)` in favor of a new option `Cert Certificate`

This is required to make the transport cacheable, since functions in golang are not comparable.
```

### CODE REVIEW

**Comment by aojea** [comment]
/assign @liggitt 

can we confirm that interfaces can be used as keys?

as you can see that makes the change backwards compatible

 I still have to understand better the integration test, it says the transport is cached but I see as much dials as object types

File: staging/src/k8s.io/client-go/rest/request_test.go [hunk 1]
```diff
@@ -3759,6 +3759,110 @@ func TestHTTP1DoNotReuseRequestAfterTimeout(t *testing.T) {
...
+	return (&net.Dialer{}).DialContext(ctx, network, address)
+}
+
+func (d *fakeDialerWithCounter) calls() int {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return d.counter
+}
+
+func TestDifferentDialers(t *testing.T) {
```
Comment (aojea) [inline staging/src/k8s.io/client-go/rest/request_test.go:3780]: Custom dialer are now cached when passed as Interface

File: staging/src/k8s.io/client-go/rest/config.go [hunk 2]
```diff
@@ -131,20 +131,42 @@ type Config struct {
 	Timeout time.Duration
 
 	// Dial specifies the dial function for creating unencrypted TCP connections.
+	// Deprecated: Use Dialer instead, which allows the transport to be cached.
+	// If both are set, Dialer takes priority.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/config.go:135]: I don't think we should allow both being set if we can't introspect `Dialer` and make sure it matches `Dial` ... setting both is a recipe for confusion/error, in my opinion

Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/config.go:135]: same comment applies to the other dual interface fields

Comment (aojea) [reply staging/src/k8s.io/client-go/rest/config.go:135]: /hold

that is game over here :smile:

Comment (aojea) [reply staging/src/k8s.io/client-go/rest/config.go:135]: /close

**Pull request closed without merging**